libm tests: split large arrays into ~4k chunks

pjonsson opened the pull request:
Some arrays have more members than what
can be supported on some targets. Split
the arrays into chunks of ~4k, and compile
a separate test for each chunk.

Refs #540

keith-packard commented on newlib/libm/test/meson.build:
```diff
@@ -157,27 +164,34 @@ foreach target : targets
-       depends: bios_bin,
-       env: test_env)
+  foreach i : range(max_test_part)
+    test_part = i.to_string()
```
A few of the run scripts rely on the `target` piece being at the end of the executable name. That means pulling the computation of `test_name` inside this loop and placing the `test_part` after `'math_test_'` and before `target`.

pjonsson replied:
Fixed.

keith-packard reviewed (changes requested):
Please fix the test name computation so that the `target` piece is at the end to keep some of the execution scripts working.

keith-packard commented:
Oh, looks like the 'range' operator is a new(ish) meson feature. We're trying to keep compatibility with some older ubuntu systems for now, can you just replace that with a list of strings?

pjonsson commented:
I'm getting consistent failures on several targets for math_test_13 when running `.github/do-linux -Dformat-default=integer -Dfreestanding=true -Dposix-io=false -Dfortify-source=2`:

```
strtod:3686, bad errno answer: (0 should be 34)
strtod:3687, bad errno answer: (0 should be 34)
```
where the numbers in question are:
```
{__LINE__, "4.0e-324", 4.0e-324, 8, },
{__LINE__, "6.0e-324", 6.0e-324, 8, },
```
I think the direct assignment to `errno` (without `_REENT_ERRNO()` wrapping) in newlib/libc/stdlib/strtod.c:339 looks suspicious, but I'm not even sure that is the strtod implementation being used for the failing tests.

I'm also puzzled why splitting the arrays would cause this failure, the failing numbers must have been tested before the split as well.

keith-packard commented:
> Edit: 34 is ERANGE, which seems right.
> 
> I'm getting consistent failures on several targets for math_test_13 when running `.github/do-linux -Dformat-default=integer -Dfreestanding=true -Dposix-io=false -Dfortify-source=2`:
> 
> ```
> strtod:3686, bad errno answer: (0 should be 34)
> strtod:3687, bad errno answer: (0 should be 34)
> ```

Well, that's an actual bug. The question isn't why this fails for you but why it works without your patch.

> I think the direct assignment to `errno` (without `_REENT_ERRNO()` wrapping) in newlib/libc/stdlib/strtod.c:339 looks suspicious, but I'm not even sure that is the strtod implementation being used for the failing tests.

Using errno directly is OK in picolibc -- there's no struct reent here, and any thread-relative location is handled by the errno macro itself.

> I'm also puzzled why splitting the arrays would cause this failure, the failing numbers must have been tested before the split as well.

Because there are a bunch of empty tests in conv_vec.c which caused the test code to stop before it reached the end of the array. -- we were missing a *lot* of tests. Splitting the array up meant starting after those entries and catching these bugs. 

Good work!

pjonsson commented:
> Because there are a bunch of empty tests in conv_vec.c which caused the test code to stop before it reached the end of the array. -- we were missing a _lot_ of tests. Splitting the array up meant starting after those entries and catching these bugs.

It's not very performant, but would address sanitizer/Valgrind have caught the bug?

keith-packard commented:
> > Because there are a bunch of empty tests in conv_vec.c which caused the test code to stop before it reached the end of the array. -- we were missing a _lot_ of tests. Splitting the array up meant starting after those entries and catching these bugs.
> 
> It's not very performant, but would address sanitizer/Valgrind have caught the bug?

Nope, there's nothing invalid with having an array element be initialized with zeros...

pjonsson commented:
This PR makes all the object files for the libm tests build on MSP430.

Making things link on MSP430 still requires some work, but the current issues with that are in the build system/build configuration, not the C sources.

pjonsson commented:
I believe I have made the requested changes, not sure if there is something I should click to indicate that.

keith-packard commented:
> I believe I have made the requested changes, not sure if there is something I should click to indicate that.

yeah, I always have trouble figuring out why github thinks this. Afraid I'll be afk until 9/5, so I may be slow to take a look at PRs.

pjonsson commented:
I think I asked while you were afk, but I believe I have addressed your comments, is there anything more that should be done to this PR?

This split makes the object files compile on msp430, and the target that has a compiler that defines `__PCCNECV70__` (comments talk about "Nec assembler") will now get validated with all the test data.

There are still a lot of link failures on msp430 relating to both code and data sizes, but being able to build the object files without scary warnings is a good first step.

keith-packard commented:
After experimenting with what this does to the arm build, I hacked it up so that splitting the tests was controlled by the 'split-large-tests' option -- that way we don't end up increasing the build and test time for existing architectures.
Take a look at https://github.com/picolibc/picolibc/tree/tests-split-large-arrays and see what you think.

keith-packard commented:
And, before we merge this, it would be good to finish the work of splitting these tests into small enough pieces to actually link on the msp430 target; I think that's just a matter of splitting the math tests across functions (so that math_test_0 only has a few of the target functions instead of all of them, for instance). What's the maximum size executable the msp430 can handle?

pjonsson commented:
> Take a look at https://github.com/picolibc/picolibc/tree/tests-split-large-arrays and see what you think.

As a whole, your patch is better, except that math.c stops running the atan2-tests for the non-split version.

pjonsson commented:
> What's the maximum size executable the msp430 can handle?

The Tmote sky is an msp430f1611 which has 10k ram and 48k rom, the Zolertia Z1 is an msp430f2617 which has 8k ram and 92k rom.

The "MSP430X" models have 20 bit pointers and size_t, but I'm looking at picolibc because newlib that comes from TI is too big for squeezing everything in on Sky/Z1. Even ignoring that, the simulator I use has limited support for the "X" instructions, so I'm not so inclined to also hit those issues when trying to upgrade the toolchain.

pjonsson commented:
I know that Github notifications aren't predictable with edits, so I'll put this in a separate comment:

Since tests can be split/not split, the run-arm script update also needs to include the previous form in the pattern match.

On a functional level, it would be nice if compiling for msp430 (and other small targets) would default to splitting tests, but I have no idea how to implement an inverted default for a selection of targets in Meson. That's more of a nice to have though, so don't let it prevent you from merging.

I will be traveling most of next week, so unlikely to respond here during that time. Will go through things and respond the week after.

keith-packard commented:
I've stuck the fixes to split the math tests into msp430-sized bites in the same branch as the rest of the hacks needed to get tests running: https://github.com/picolibc/picolibc/tree/test-msp430

keith-packard commented:
oh, this also leaves the tests all mashed together on the 'big' platforms.

pjonsson commented:
> I've stuck the fixes to split the math tests into msp430-sized bites in the same branch as the rest of the hacks needed to get tests running: https://github.com/picolibc/picolibc/tree/test-msp430

I've added some comments on the commits in that branch. The clean parts have already been merged to main, so your hacks look similar to mine. I also have a bunch of msp430-specific things in `picolibc.ld.in` that I haven't figured out how to handle, but I figured I could wrestle Meson after things were running.

keith-packard commented:
Sounds good. I've got tests linking and running, but the linker script is definitely not optimal. I suspect applications will need to customize it as you probably want to stick some code below the 16-bit boundary, but the interrupt vector sits right at the end in a fixed location making it hard to arrange things correctly. I've got a kludge which places the _start function at 0xff00, the interrupt vector at 0xff80 and all of the rest of the code above 0x10000.

pjonsson commented:
I was going to ask if your branch isn't based off latest main, but it turns out I forgot to make a second pull request for qsort.c, sorry about that. Your stack fixes and sizeof-fixes are improvements, but I think it's easier to just switch to C99 types: https://github.com/pjonsson/picolibc/pull/new/tests-qsort-arrraytype

I notice you replied to the interrupt vector comment, and I guess those things can be fixed later. I feel the remaining comments in the first commit are essential to address though, otherwise they will slip through the cracks and we'll be without atan2-tests.

keith-packard commented:
Would you believe I've got all of the tests working *and passing*? Found a handful of real bugs here. There are a couple of 'questionable' patches which work around bugs in either the compiler or the simulator (my bet is the simulator).

So, this work has definitely been worth the effort; without it, there are several functions which generate incorrect results on any 16-bit int target.

pjonsson commented:
I worked a bit on newlib for a 16-bit target a few years ago, so I had a hunch about how much work picolibc would be. And while I realize this kind of work is in your area of expertise and that you have a lot of picolibc in your mental cache, you're still insanely fast.

I think you are right about where the bugs are most likely to reside, but I feel that I have repeatedly mentioned that leaning on the MSP430X support is a risk since it isn't complete.

The two work arounds don't fill my heart with joy, can the rest be merged now, and the workarounds be put on a branch for the moment? I will take a stab at them in a week.

pjonsson commented:
I had another look at the code and I realized that:
```C
static inline uint32_t floor_log2(const uint32_t value) {
#if __SIZEOF_INT__ >= 4
  return 31 - __builtin_clz(value);
#elif __SIZEOF_INT__ < 4 && __SIZEOF_LONG__ >= 4
  return 31 - __builtin_clzl(value);
#else
#error no usable clz
#endif
}
```
will fail on targets with a byte size > 8 bits. Changing the test to:
```C
static inline uint32_t floor_log2(const uint32_t value) {
#if INT_MAX >= INT32_MAX
  return 31 - __builtin_clz(value);
#elif INT_MAX < INT32_MAX && LONG_MAX >= INT32_MAX
  return 31 - __builtin_clzl(value);
#else
#error no usable clz
#endif
}
```
should make it work as intended.

keith-packard commented:
Yeah, I suppose. Hard to imagine we'll ever find a target for picolibc that doesn't use 8-bit bytes though :-).

pjonsson commented:
The non-8-bit-byte isn't something that needs to be fixed right now, but can I make a PR that fixes that in a couple of weeks?

keith-packard commented:
> The non-8-bit-byte isn't something that needs to be fixed right now, but can I make a PR that fixes that in a couple of weeks?

of course.

pjonsson commented:
This is superseded by #582.
